### ISSUE TITLE
Export API-Details for Post-Processing

### ISSUE THREAD

**Issue by cwiechmann** [Axway-API-Management-Plus/apimanager-swagger-promote#60]
If I want to do some post processing after I run the script, what are the ways? It would be nice if script return the generated API-ID for post-processing. 
The scenario:
I create config file on the fly based on input parameters. If I want to update description markdown file path, I can do in config file at API level. But at the method description markdown file can be given after the API is created. I can use rest call to do. But how do i know the apiId that swagger script created? I can query based on api name using rest call but I am not comfortable as there may be multiple apis with same names .

**Comment by cwiechmann** [comment]
I'm thinking about an option where you can provide a properties-file, which will contain key-value pairs about details of the processing. 
Option let's say: -e myGeneratedAPI.properties

Initially this file should contain something like:
feApiId=dfs3-32312-dfasd312-3123-1
beApiID=dadas3-3231

And so forth. This could be easily extended in future releases about additional information.

Comments on this are welcome.

**Comment by hakila** [comment]
I am already doing that. I wrote a script that takes parameters like path, api name, image , tags etc from input file and generates config file on the fly and then provide this config file to your swagger promote script. They all are key value pairs in my prop file. Are you saying same? Correct me if I am wrong.

**Comment by cwiechmann** [comment]
The idea is more that Swagger-Promote creates that file for you, after the API has been imported into the API-Manager. You would instruct Swagger-Promote to do this, by providing this new mentioned parameter (-e myGeneratedAPI.properties).  
A Post-Processing-Tool could now read that file, gets the actual API-ID (and maybe other information in the future) and can do whatever is needed to do.

**Comment by hakila** [comment]
That sound good. How soon this enhancement will be available?

**Comment by cwiechmann** [comment]
Not sure, when this is added. It depends a bit on demand.  
First I would like to finally implement the Export feature, which can be used to export the required artifacts (API-Config, Swagger-File, Image, ...) from a running API-Manager and maybe Re-Import it into another API-Manager.

**Comment by cwiechmann** [comment]
Released with version 1.5.3

**Comment by hakila** [comment]
I am invoking script as below
/opt/axway/scripts/bin/APIcreate.py , where in this python I am invoking swaggerrpomote script
 os.system(SWAGGER_IMPORT_SCRIPT + " -u " + USER_ID + " -p "+ PWD+ " -a " + "'" + apiCreateJsonFileNameArg  + "'" + " -c "+ "'" + apiConfigJsonFileNameArg + "'" + ' -f true' + ' -e propgen.properties ')
Api is created but property file is not generated at /opt/axway/scripts/bin. Please suggest where I am doing wrong or example on how to use this feature

**Comment by cwiechmann** [comment]
Using parameter -e was the idea as I responded to that issue, but the implementation is using the following:
`-detailsExportFile <APIDetails.properties>   Configure a filename, to get a Key=Value file containing information about the created API.`

Just tested it locally and you should see the following line logged:
`Created API-Properties file: 'C:\temp\swagger-promote-1.5.3\samples\APIDetails.properties'`

The file is created relative to the API-Config-File or using an absolute path:
`-detailsExportFile C:/temp/APIDetails.properties`
`Created API-Properties file: 'C:\temp\APIDetails.properties'`

**Comment by hakila** [comment]
Thank you for the info. Where can I find all these details? Do we have detailed release notes?

**Comment by cwiechmann** [comment]
There just the release-notes in the "Releases" section. Swagger-Promote provides a usage, which is shown when calling it without any arguments.

**Comment by hakila** [comment]
For post processing, i want to make change only if the api is changed. Currently script return 0 for both changed and unchanged. Is is possible to return different return code if it is unchnaged or some mechanism.

**Comment by cwiechmann** [comment]
The Java-Part already returns a dedicated return-code for a No-Change, which is just handled by the run-swagger-script.
Have a look into that script and you can change it yourself.
‘
if [ $rc -eq 10 ];then
        echo "No changes detected. Existing with RC: 0"
        exit 0
fi
‘

**Comment by hakila** [comment]
Thank you for quick response.

**Comment by hakila** [comment]
I was testing this. It seems you are returning apiId as frontendId. I am doing rest call and it is failing
https://devaxwaypoc01.atpco.org:8075/api/portal/v1.3/proxies/611471c0-1082-4414-a3cd-59b47a6d6491/operations
<Response [403]>
Generated prop file contained this
feApiId=611471c0-1082-4414-a3cd-59b47a6d6491

{
        "id": "0a3cd1df-ab14-4c49-a81b-60bf6730408a",
        "organizationId": "33bff549-c285-46bc-911f-60dfa105b006",
        "apiId": "611471c0-1082-4414-a3cd-59b47a6d6491",
        "name": "TEST SWGIMPSCRPT16SS Minimum Stay (Category 06) Rules",
        "version": null,
        "apiRoutingKey": null,
        "vhost": null,
        "path": "/services/testswgimpscrpt16ss/v1/rules/record3/cat6",
        "descriptionType": "markdown",
        "descriptionManual": null,
        "descriptionMarkdown": "${env.APIDOCS}/markdown/seasonalitycat3footnotes/portal/seasonalitycat3footnotes.md",
        "descriptionUrl": null,
        "summary": null,
        "retired": false,
        "expired": false,
        "image": "/api/portal/v1.3/discovery/swagger/apis/0a3cd1df-ab14-4c49-a81b-60bf6730408a/image",
        "retirementDate": 0,
        "deprecated": false,
        "state": "published",
        "createdOn": 1563219255111,
        "createdBy": "ac092774-ad5f-44aa-8db2-fa55cf4b17d2",
        "corsProfiles": [
            {
                "name": "_default",
                "isDefault": true,
                "origins": [
                    "*"
                ],
                "allowedHeaders": [],
                "exposedHeaders": [
                    "X-CorrelationID"
                ],
                "supportCredentials": false,
                "maxAgeSeconds": 0
            }
        ]
Please investigate.

**Comment by cwiechmann** [comment]
Yes, you are right. I made a mistake by using getApiId() instead of getId() :astonished: in 
`com.axway.apim.actions.CreateNewAPI`
line:
`APIPropertiesExport.getInstance().setProperty("feApiId", createdAPI.getApiId());`

I’m off for the following weeks. Maybe you can fix it yourself and build it with 
`mvn clean package -DskipTests` 
This gives you a fresh build in the distribution module.
And of course I’m happy to merge a Pull-Request for that fix.

**Comment by hakila** [comment]
seems, I dont have access. 
![gitcommiterr](https://user-images.githubusercontent.com/17910073/61327523-14d07300-a7e7-11e9-901a-a45653dcd5d4.png)

**Comment by cwiechmann** [comment]
Yes, that’s correct. The idea is to use Pull-Reqests based on a fork you have created.
See here how it works:
https://help.github.com/en/articles/creating-a-pull-request-from-a-fork

**Comment by hakila** [comment]
I created fork and able to commit my change. Please do Merge when you get a chance.

**Comment by hakila** [comment]
I made change and testing with locally generated swagger promote. I am still seeing backend api coming in property file generated. COuld you please check on your end?

**Comment by hakila** [comment]
API-Manager Promote Version: 1.6.1-SNAPSHOT

To report issues or get help, please visit:
https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote
------------------------------------------------------------------------

0      INFO  nvironmentProperties| Loaded environment properties from file: env.properties.
797    INFO     APIManagerAdapter| API-Manager version is: 7.7.0
920    INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: '/opt/axway/pythonscripts/conf/apicreatejson/TESTSWGIMSCRT16SSMinimumStayCategory06Rules_transformed.json' (relative path)
922    INFO  IImportConfigAdapter| Handling configured client-applications.
986    INFO  IImportConfigAdapter| Found existing application: 'ATP_ATP_Apps' based on given name 'ATP_ATP_Apps'
1080   INFO     APIManagerAdapter| Found existing API on path: '/services/poc/v1/rules/record3/cat6' (published) (ID: '9e5981e3-fb26-40ff-bb1a-e1fff81e0b8f')
2248   INFO        APIChangeState| Changed property: image[Desired: 'APIImage [bytes=9654]' vs Actual: 'APIImage [bytes=13883]']
2248   INFO     APIManagerAdapter| Strategy: Going to update existing API: TEST SWGIMSCRT16SS Minimum Stay (Category 06) Rules (Version: null)
2248   INFO     APIManagerAdapter| Recognized the following changes. Potentially Breaking: [] plus Non-Breaking: [image]
2248   INFO     APIManagerAdapter| Strategy: Update existing API, as all changes can be applied in current state.
2288   INFO     APIManagerAdapter| Updating API-Image from: /opt/axway/apidocs/markdown/config/resources/images//Ticket.png
2326   INFO     APIManagerAdapter| Default-Application-Quota for API: 'TEST SWGIMSCRT16SS Minimum Stay (Category 06) Rules' is UN-CHANGED. Nothing to do.
2327   INFO     APIManagerAdapter| All desired organizations: [ATP_ATP_Org, API Development] have already access. Nothing to do.
2328   INFO     APIManagerAdapter| All desired applications: [[ATP_ATP_Apps]] have already a subscription. Nothing to do.
2331   INFO                   App| Created API-Properties file: '/opt/axway/pythonscripts/conf/apiconfigjson/TESTSWGIMSCRT16SSMinimumStayCategory06RulesDetails.prop'
2331   INFO                   App| Successfully replicated API-State into API-Manager
retCode from swagger script: 0
apiFrontendId: 7e38a75b-078e-4efb-aecc-05b6e1848edb
https://devaxwaypoc01.atpco.org:8075/api/portal/v1.3/proxies/7e38a75b-078e-4efb-aecc-05b6e1848edb/operations
<Response [403]>

Above 7e38a75b-078e-4efb-aecc-05b6e1848edb is the backend id but not the frontend id.

**Comment by hakila** [comment]
Actually the above api existed before and i tested this new swagger(locally built) against that. Then I did not see the change reflected. But later I deleted api manually and tested again with new swagger(locally built). Then it is working
------------------------------------------------------------------------
API-Manager Promote Version: 1.6.1-SNAPSHOT

To report issues or get help, please visit:
https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote
------------------------------------------------------------------------

0      INFO  nvironmentProperties| Loaded environment properties from file: env.properties.
791    INFO     APIManagerAdapter| API-Manager version is: 7.7.0
893    INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: '/opt/axway/pythonscripts/conf/apicreatejson/TESTSWGIMSCRT16SSMinimumStayCategory06Rules_transformed.json' (relative path)
894    INFO  IImportConfigAdapter| Handling configured client-applications.
956    INFO  IImportConfigAdapter| Found existing application: 'ATP_ATP_Apps' based on given name 'ATP_ATP_Apps'
1053   INFO     APIManagerAdapter| No existing API found exposed on: '/services/poc/v1/rules/record3/cat6'
1054   INFO     APIManagerAdapter| Strategy: No existing API found, creating new!
1056   INFO     APIManagerAdapter| Importing backend API (Swagger/WSDL Import)
1145   INFO     APIManagerAdapter| Create API-Proxy (Front-End API)
2161   INFO     APIManagerAdapter| Updating proxy for the following properties: descriptionType descriptionMarkdown securityProfiles tags outboundProfiles corsProfiles path name
2183   INFO     APIManagerAdapter| Updating API-Image from: /opt/axway/apidocs/markdown/config/resources/images//Ticket.png
2237   INFO     APIManagerAdapter| Updating Application-Default-Quota for API: TEST SWGIMSCRT16SS Minimum Stay (Category 06) Rules
2387   INFO     APIManagerAdapter| Application-Default quota successfully updated: TEST SWGIMSCRT16SS Minimum Stay (Category 06) Rules
2410   INFO     APIManagerAdapter| Granted permission to organization: '[ATP_ATP_Org]'
2411   INFO     APIManagerAdapter| Creating API-Access for the following apps: '[[ATP_ATP_Apps]]'
2438   INFO                   App| Created API-Properties file: '/opt/axway/pythonscripts/conf/apiconfigjson/TESTSWGIMSCRT16SSMinimumStayCategory06RulesDetails.prop'
2438   INFO                   App| Successfully replicated API-State into API-Manager
retCode from swagger script: 0
apiFrontendId: dfd4b497-a606-47af-b740-791f7da65be2
https://devaxwaypoc01.atpco.org:8075/api/portal/v1.3/proxies/dfd4b497-a606-47af-b740-791f7da65be2/operations
<Response [200]>

In both cases api got created but why it did not work with existing api?

**Comment by hakila** [comment]
Change has to be made in UpdateExistingAPI.java as well. I just changed and tested and it is working.
I also committed my change. Please merge when you get a chance.

**Comment by hakila** [comment]
I am testing orgs. I had org ggg that was not existed. And I removed  org that api had before.
0      INFO  nvironmentProperties| Loaded environment properties from file: env.properties.
815    INFO     APIManagerAdapter| API-Manager version is: 7.7.0
930    INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: '/opt/axway/pythonscripts/conf/apiswagger/apicreatejson/TESTSWGIMSCRT16SSMinimumStayCategory06Rules_transformed.json' (relative path)
932    INFO  IImportConfigAdapter| Handling configured client-applications.
1004   INFO  IImportConfigAdapter| Found existing application: 'ATP_ATP_Apps' based on given name 'ATP_ATP_Apps'
1004   INFO  IImportConfigAdapter| Found existing application: 'ATP_CUS_App' based on given name 'ATP_CUS_App'
1004   INFO  IImportConfigAdapter| Found existing application: 'X91_91_App' based on given name 'X91_91_App'
1106   INFO     APIManagerAdapter| Found existing API on path: '/services/poc/v1/rules/record3/cat6' (published) (ID: '5a819904-3be4-4d20-a361-67b24cc6d430')
2181   INFO        APIChangeState| Changed property: clientOrganizations[Desired: '[ATP_ATP_Org, X91_91_Org, gggg, API Development]' vs Actual: '[ATP_CUS_Org, API Development, ATP_ATP_Org, X91_91_Org]']
2328   INFO        APIChangeState| Changed property: image[Desired: 'APIImage [bytes=9654]' vs Actual: 'APIImage [bytes=13883]']
2328   INFO     APIManagerAdapter| Strategy: Going to update existing API: TEST SWGIMSCRT16SS Minimum Stay (Category 06) Rules (Version: null)
2328   INFO     APIManagerAdapter| Recognized the following changes. Potentially Breaking: [] plus Non-Breaking: [clientOrganizations, image]
2328   INFO     APIManagerAdapter| Strategy: Update existing API, as all changes can be applied in current state.
2372   INFO     APIManagerAdapter| Updating API-Image from: /opt/axway/apidocs/markdown/config/resources/images//Ticket.png
2411   INFO     APIManagerAdapter| Default-Application-Quota for API: 'TEST SWGIMSCRT16SS Minimum Stay (Category 06) Rules' is UN-CHANGED. Nothing to do.
2411   ERROR    APIManagerAdapter| Requested OrgId for unknown orgName: gggg
2411   ERROR    APIManagerAdapter| Configured organizations: ['XBW_BW_Org', 'XAT_AB_Org', 'XUA_UA_ORG', 'V1S_EY_Org', 'W530_BA_Org', 'W530_91_Org', 'W530_VS_Org', 'XUA_UA_Org', 'XLA_LA_Org', 'XDL_DL_Org', 'XKL_KL_Org', 'W055_055_Org', 'XAB_AB_Org', 'ATP_SWT_ORG', 'XJJ_JJ_Org', 'V82_82_Org', 'yu3r3f', 'XWS_WS_ORG', 'dsd', 'W135_135_Org', 'a1Org', 'ATP_CUS_Org', 'W530_IB_Org', 'WFZ_FZ_ORG', 'W123_123_Org', 'VIS_EY_Org', 'XBA_BA_Org', 'W530_AM_Org', 'API Development', 'XTP_TP_Org', 'X15_15_Org', 'W300_300_Org', 'W530_MU_Org', 'XTA_TA_Org', 'XPZ_PZ_Org', 'ATP_ATP_Org', 'test', 'V82_91_Org', 'ATP_MON_ORG', 'XMA_MA_Org', 'XUS_US_Org', 'XEY_EY_Org', 'Community', 'W530_DL_Org', 'XBRL_DUM_Org', 'XAC_AC_Org', 'W119_119_Org', 'XWS_WS_Org', 'XXL_XL_Org', 'XAF_AF_Org', 'XAA_AA_Org', 'W109_109_Org', 'INTERES_OPEN_ORG', 'X91_91_Org', 'W530_EY_Org', 'XFC_FC_Org']
2412   ERROR                  App| Unknown Org-Name: 'gggg'
retCode from swagger script: 0
 If you see the desired vs actual, script did not do anything for ATP_CUS_Org due to the error for gggg which was not existed. I expect script return error code other than 0. I am doing post processing using the feapid based on error code. In this case property file that contain feapiid is not generated.  Please suggest.

**Comment by cwiechmann** [comment]
@hakila, for that error situation the Return-Code 57 should be returned and I see no reason why it shouldn't. There is even an Integration-Tests making sure it returns with code: 57.

I tested the behavior by registering an API with 3 Client-Orgs:
```
0      INFO  nvironmentProperties| Loaded environment properties from file: env.properties.
2537   INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: 'samples/petstore.json' (relative path)
2556   INFO     APIManagerAdapter| API-Manager version is: 7.7.0
2671   INFO     APIManagerAdapter| No existing API found exposed on: '/sample/minimal/apidefinition/api/v1'
2673   INFO     APIManagerAdapter| Strategy: No existing API found, creating new!
2674   INFO     APIManagerAdapter| Importing backend API (Swagger/WSDL Import)
3017   INFO     APIManagerAdapter| Create API-Proxy (Front-End API)
5508   INFO     APIManagerAdapter| Updating proxy for the following properties: securityProfiles corsProfiles path version name
5640   INFO     APIManagerAdapter| Granted permission to organization: '[API Development 1013, API Development 1119, Chris Test]'
5646   INFO                   App| Created API-Properties file: 'C:\temp\swagger-promote-1.6.1-SNAPSHOT\samples\basic\exportPropeties.txt'
5647   INFO                   App| Successfully replicated API-State into API-Manager
```
For that initial load the return-code:
```
C:\temp\swagger-promote-1.6.1-SNAPSHOT>echo Exit Code is %errorlevel%
Exit Code is 0
```

After that I removed the organization "Chris Test" from the API-Manager and re-run Swagger-Promote:
```
0      INFO  nvironmentProperties| Loaded environment properties from file: env.properties.
2455   INFO  IImportConfigAdapter| Reading API-Definition (Swagger/WSDL) from file: 'samples/petstore.json' (relative path)
2473   INFO     APIManagerAdapter| API-Manager version is: 7.7.0
2584   INFO     APIManagerAdapter| Found existing API on path: '/sample/minimal/apidefinition/api/v1' (published) (ID: '0f1b3718-479e-4c68-8669-af3d95a81ffa')
3819   INFO        APIChangeState| Changed property: clientOrganizations[Desired: '[API Development 1013, API Development 1119, Chris Test, API Development]' vs Actual: '[API Development 1013, API Development, API Development 1119]']
3819   INFO     APIManagerAdapter| Strategy: Going to update existing API: Sample Minimal API (Version: 1.0.0)
3819   INFO     APIManagerAdapter| Recognized the following changes. Potentially Breaking: [] plus Non-Breaking: [clientOrganizations]
3819   INFO     APIManagerAdapter| Strategy: Update existing API, as all changes can be applied in current state.
3876   ERROR    APIManagerAdapter| Requested OrgId for unknown orgName: Chris Test
3877   ERROR    APIManagerAdapter| Configured organizations: ['API Test-Org 3 2632', 'API Development 8346', 'API Development 5935', 'API Development 9108', 'API Test-Org 3 4166', 'Partners', 'API Test-Org 3 2076', 'API Test-Org 2 7588', 'API Development 3932', 'API Development 9278', 'API Test-Org 3 5495', 'API Development 6696', 'API Development 5989', 'API Test-Org 2 4166', 'API Development 8638', 'API Development 3103', 'API Development 2150', 'API Development 7911', 'API Development 9306', 'API Test-Org 3 5036', 'API Development 1183', 'API Test-Org 3 2304', 'API Development 7750', 'API Test-Org 2 3457', 'API Test-Org 3 8136', 'API Test-Org 2 5036', 'Plexus Health', 'API Test-Org 2 5495', 'API Test-Org 2 9339', 'API Test-Org 2 3570', 'API Test-Org 3 9340', 'API Test-Org 2 2304', 'API Test-Org 2 4036', 'FastCars', 'API Test-Org 2 9340', 'API Test-Org 2 2076', 'API Development 6043', 'API Test-Org 2 5651', 'API Test-Org 3 5651', 'API Test-Org 3 6426', 'Community', 'API Test-Org 3 3641', 'API Development 2921', 'API Test-Org 3 3946', 'API Development 1013', 'API Test-Org 2 8136', 'API Test-Org 3 9339', 'API Test-Org 2 2632', 'API Development 4917', 'API Test-Org 3 7045', 'API Test-Org 2 3641', 'API Test-Org 2 9255', 'API Test-Org 2 4516', 'API Test-Org 2 7045', 'API Test-Org 3 3191', 'API Test-Org 3 7588', 'API Development', 'API Test-Org 2 4006', 'API Development 8595', 'API Development 8601', 'FHIR', 'API Development 1655', 'API Test-Org 3 8037', 'API Test-Org 2 1826', 'API Test-Org 3 4036', 'API Development 9697', 'API Development 1119', 'API Test-Org 3 1826', 'API Test-Org 3 3570', 'API Test-Org 2 8925', 'API Development 1751', 'API Test-Org 3 4516', 'API Test-Org 2 5741', 'API Development 2973', 'API Test-Org 3 4006', 'API Development 6753', 'API Test-Org 3 8925', 'API Test-Org 2 8037', 'API Test-Org 2 6426', 'API Test-Org 3 3457', 'API Test-Org 2 3191', 'API Test-Org 3 5741', 'API Test-Org 3 9255', 'API Test-Org 2 3946']
3881   INFO                   App| Created API-Properties file: 'C:\temp\swagger-promote-1.6.1-SNAPSHOT\samples\basic\exportPropeties2.txt'
3881   ERROR                  App| Unknown Org-Name: 'Chris Test'
```
And the return code:
```
C:\temp\swagger-promote-1.6.1-SNAPSHOT>echo Exit Code is %errorlevel%
Exit Code is 57
```
:exclamation: But you are right, that the API-Post-Processing file wasn't created in that case and I fixed that now!

[apimanager-swagger-promote-1.6.1-SNAPSHOT.tar.gz](https://github.com/Axway-API-Management-Plus/apimanager-swagger-promote/files/3478645/apimanager-swagger-promote-1.6.1-SNAPSHOT.tar.gz)
